curvefs/src/mds/topology/topology.cpp

```diff
@@ -1780,6 +1785,12 @@ TopologyImpl::AllocOrGetMemcacheCluster(FsIdType fsId,
     return ret;
 }
 
+bool TopologyImpl::Tso(uint64_t* ts, uint64_t* timestamp) {
+    *timestamp = curve::common::TimeUtility::GetTimeofDayMs();
```

Comment:
This is indeed a problem. If the subsequent request obtains a smaller timestamp, it may roll back the ongoing transaction in a transaction conflict scenario. However, this requires a logically monotonically increasing time to confirm whether the transaction has timed out. Do you have any feasible suggestions?

Reply:
Can try this function [clock_gettime](https://linux.die.net/man/3/clock_gettime)? `CLOCK_MONOTONIC` args seems can do it. can `man clock_gettime` see details.

Reply:
`clock_gettime` with args `CLOCK_REALTIME ` is better than `gettimeofday` here but the clocks on different MDS machines may still have discrepancies.

Reply:
> Can try this function [clock_gettime](https://linux.die.net/man/3/clock_gettime)? `CLOCK_MONOTONIC` args seems can do it. can `man clock_gettime` see details.

`clock_gettime` with args `CLOCK_REALTIME ` is better than `gettimeofday` here but the clocks on different MDS machines may still have discrepancies.

Reply:
> `clock_gettime` with args `CLOCK_REALTIME ` is better than `gettimeofday` here but the clocks on different MDS machines may still have discrepancies.

What I mean is that this function can get continuous and greater than over and over timestamps on a single machine, and is not affected by time drift and NTP alignment. If you want to get this on multiple machines, It's going to be a big challenge. For most storage or data services, they have a separate timing service.